starlite/kwargs.py

```diff
@@ -432,16 +510,14 @@ async def resolve_dependency(
             The resolved dependency value
         """
         if dependency.provide.cache_per_request:
-            async with dependency.provide.lock:
```

Comment:
Without the lock the `cache_per_request` config won't work as expected (at least the way I imagined it).
Since I ended up not needing this, we could just remove that functionality completely.

Coming from FastAPI, the current default may be confusing though. Their behavior is to cache dependencies per request, but resolve them sequentially. This way they don't need a lock, but if someone has multiple async dependencies they won't run concurrently.

Maybe a better way would be to collect all required dependencies first, batch them into groups based on their sub-dependencies and then sequentially resolve the batches?
An example:
Dependencies A and B have no dependencies, C and D both require A, E requires B and C.
Resolve A and B in the first batch concurrently, then resolve C and D in the second batch (also  concurrently) and finally resolve E.
B could also be resolved in the second batch, which would require one less TaskGroup.

Reply:
so the issue why I removed this - locks are apparently extremely memory intensive to create and save. We need to come up with a better strategy to handle this that does not impact resources like this. Whatever we gain from async resolution using a task group, we are losing form the overhead of this.

Reply:
@Goldziher can we do concurrent resolution ONLY if user explicitly requests it?

like xpresso does - https://xpresso-api.dev/0.42.3/advanced/dependencies/performance/#concurrent-execution

Reply:
@Goldziher I reworked the way dependencies are resolved as per my suggestion above: https://github.com/ottermata/starlite/commit/dd229a0d06d4a3289319631c141899ab74afaea4

The main difference with this approach is that all dependencies only ever get called once per request. I can't really imagine a reason why someone would want to have them executed multiple times, so I think this is fine.

Can you run you benchmarks with that change included? I can open a PR if needed.

Reply:
Sure, open a pr

Reply:
@ottermata Can you rebase against the `performance_updates` branch in a separate branch of yours? Then I can benchmark the three.

Reply:
@provinzkraut Here's the branch: https://github.com/ottermata/starlite/tree/performance_updates/rework_dependencies

Reply:
@provinzkraut I rebased against the current version